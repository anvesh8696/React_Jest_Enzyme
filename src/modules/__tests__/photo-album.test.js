import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';
import PhotoAlbum from '../photo-album';
import fakePhotos from '../__data__/photos.fake.js';


describe('<PhotoAlbum />', () => {
	let photoAlbum,
	fakePromise,
	wrapper,
	fakePromiseResolve,
	fakeResponse,
	sandbox,
	chancee,
	numPhotos;
	const rows = 3;
	const cols = 3;

	const mockResponse = (status, statusText, response) => {
	  return new window.Response(response, {
	    status: status,
	    statusText: statusText,
	    headers: {
	      'Content-type': 'application/json'
	    }
	  });
	};

	const renderPhotoAlbum = function(){
		wrapper = shallow(<PhotoAlbum rows={rows} cols={cols}	 />);
		photoAlbum = wrapper.instance();
	}

	const setupPromise = function(){
		return new Promise((resolve, reject) => {
			fakePromiseResolve = resolve;
		});
	}

	const itAsync = (description, func) => {
	    it(description, async (done) => {
	        try {
		    await func();
		    done();
	        } catch (err) {
		    done.fail(err);
	        }
	    });
	};

	const afterPromises = (done, fn) => {
	  setTimeout(() => {
	    try {
	      fn();
	      done();
	    } catch(e) {
	      done.fail(e);
	    }
	  }, 0);
	}


	beforeEach(() =>{
		renderPhotoAlbum();
		fakeResponse = mockResponse(200, 'OK', JSON.stringify(fakePhotos));
		fakePromise = setupPromise();
		global.fetch = jest.fn().mockImplementation(() => fakePromise);
		photoAlbum.componentDidMount();
	});


	it('fetch was called', () => {
		expect(global.fetch).toHaveBeenCalledTimes(1);
		expect(global.fetch).toHaveBeenCalledWith("https://jsonplaceholder.typicode.com/photos", {});
	});


	it("state was populated with photos", (done) => {
		numPhotos = cols * rows;
		afterPromises(done, () => {
			expect(photoAlbum.state.photos.length).toEqual(numPhotos);
			expect(photoAlbum.state.photos).toEqual(fakePhotos);
		});
		fakePromiseResolve(fakeResponse);
	});

	it('matches snapshot', () => {
		expect(toJson(wrapper)).toMatchSnapshot();
	});


});

