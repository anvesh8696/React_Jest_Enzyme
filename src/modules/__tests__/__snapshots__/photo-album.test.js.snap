exports[`<PhotoAlbum /> matches snapshot 1`] = `
<div>
  <h1>
    Photo Album
  </h1>
  <Thumbnails
    cols={3}
    photos={Array []}
    rows={3}
    total={9} />
</div>
`;
